text/0000-futures.md

```diff
@@ -0,0 +1,520 @@
+`LocalWaker` to provide more effenciency when the wakeup is guaranteed to occur on the
+executor thread:
+
+```rust
+/// A `Waker` is a handle for waking up a task by notifying its executor that it
+/// is ready to be run.
+///
+/// This handle contains a trait object pointing to an instance of the `UnsafeWake`
+/// trait, allowing notifications to get routed through it.
+///
+/// Implements `Clone`, `Send`, and `Sync`.
+pub struct Waker { ... }
+
+impl Waker {
+    /// Wake up the task associated with this `Waker`.
+    pub fn wake(&self);
+}
+
+/// A `LocalWaker` is a handle for waking up a task by notifying its executor that it is ready to be run.
+///
+/// This is similar to the `Waker` type, but cannot be sent across threads. Task executors can use this type to implement more optimized singlethreaded wakeup behavior.
+impl LocalWaker {
+    /// Wake up the task associated with this `LocalWaker`.
+    pub fn wake(&self);
+}
+
+/// You can upgrade to a sendable `Waker` at zero cost, but waking through a `Waker` is more expensive
+/// due to synchronization.
+impl From<LocalWaker> for Waker  { .. }
+```
+
+Task execution always happens in the context of a `LocalWaker` that can be used to
+wake the task up locally, or converted into a `Waker` that can be sent to other threads.
+
+It's possible to construct a `Waker` using `From<Arc<dyn Wake>>`.
+
+### `UnsafeWake` and `no_std` compatibility
+
+The [`UnsafeWake` trait](https://doc.rust-lang.org/nightly/std/task/trait.UnsafeWake.html)
+in `core::task` is designed to support task wakeup in  `no_std` environments, where
+we cannot use `Arc`. It is *not* proposed for stabilization at this time, because
+its APIs are awaiting revision based on object safety for `*mut self` methods.
+
+## `core::future` module
+
+With all of the above task infrastructure in place, defining `Future` is
+straightforward:
+
+```rust
+pub trait Future {
+    /// The type of value produced on completion.
+    type Output;
+
+    /// Attempt to resolve the future to a final value, registering
+    /// the current task for wakeup if the value is not yet available.
+    ///
+    /// # Return value
+    ///
+    /// This function returns:
+    ///
+    /// - [`Poll::Pending`] if the future is not ready yet
+    /// - [`Poll::Ready(val)`] with the result `val` of this future if it
+    ///   finished successfully.
+    ///
+    /// Once a future has finished, clients should not `poll` it again.
+    ///
+    /// When a future is not ready yet, `poll` returns `Poll::Pending` and
+    /// stores a clone of the [`LocalWaker`] to be woken once the future can
+    /// make progress. For example, a future waiting for a socket to become
+    /// readable would call `.clone()` on the [`LocalWaker`] and store it.
+    /// When a signal arrives elsewhere indicating that the socket is readable,
+    /// `[LocalWaker::wake]` is called and the socket future's task is awoken.
```

Comment:
Store it where? Called by whom?

Reply:
As [I said above](https://github.com/rust-lang/rfcs/pull/2592#discussion_r233434961), they're typically stored somewhere so that another task or an IO event loop can call `wake()` on them. For more information, I'd recommend checking out the relevant sections of the async book: 
- [Wakeups](https://github.com/rust-lang-nursery/wg-net/blob/master/async-book/src/execution/wakeups.md)
- [Executors and System IO](https://github.com/rust-lang-nursery/wg-net/blob/master/async-book/src/execution/io.md)